### PR TITLE
K-space pixel integration: sinc+wrap replacing spatial oversampling

### LINKED ISSUE #38: Decouple pixel response (oversample) from PSF configuration
## Problem

Pixel response (integrating flux over pixel area) and PSF convolution are currently coupled in the code. `build_image_obs` only activates oversampling when a PSF is configured — without a PSF, it forces `oversample=1`, which means no pixel integration.

This is physically wrong: pixel response is a detector property that's always present, regardless of whether an optical PSF is modeled. A steep profile (e.g., exponential cusp, or Spergel with nu < 0) without a PSF is exactly where pixel integration matters most, yet it's the case where the code skips it.

The coupling exists because PSF convolution and pixel integration share the oversampling machinery (render at fine resolution, convolve/bin, downsample). But they're independent physical effects that should be independently configurable.

## Current behavior

- `build_image_obs(ip, psf=psf_obj)` → `oversample=5` (pixel response ON)
- `build_image_obs(ip)` → forces `oversample=1` (pixel response OFF)
- `render_image(theta, image_pars=ip)` → `oversample=1` default (pixel response OFF)
- Synthetic data (`_generate_sersic_scipy`) also defaults to `oversample=1` (point-sampled)

This means no-PSF tests compare point-sampled model vs point-sampled data — internally consistent but both wrong in the same way.

## Desired behavior

- `oversample=5` (pixel integration) should be the default everywhere, independent of PSF
- PSF on/off is a separate toggle
- `oversample=1` should be an explicit opt-out for cases that specifically need point-sampling
- Synthetic data generation should also pixel-integrate by default

## Found during

Spergel model implementation (`se/inclined-sersic`). The Spergel profile with nu < 0 has a central cusp (divergent SB) that makes the point-sampling vs pixel-integration distinction very visible. Face-on GalSim regression tests exposed the issue: our `oversample=1` rendering and GalSim `method='no_pixel'` both point-sample, masking the fact that neither computes what a real detector would measure.

## TODO

- [ ] Change `build_image_obs` default: always `oversample=5`, independent of PSF
- [ ] Change `render_image` default: `oversample=5`
- [ ] Update `_generate_sersic_scipy` and `_generate_spergel_scipy` to default `oversample=5`
- [ ] Switch GalSim regression tests from `method='no_pixel'` to `method='auto'`
- [ ] Re-validate all existing tests with new defaults
- [ ] Update tolerance values if pixel-integrated comparisons shift residuals
- [ ] Verify likelihood slice and optimizer recovery tests still pass

### PR DESCRIPTION
**Pull request by sweverett**
## Summary

Replaces O(N^2) spatial oversampling for pixel integration with exact k-space sinc multiplication following GalSim's wrap approach. The rendering chain is now `IFFT(profile_FT * pixel_sinc * PSF_FT)` in a single FFT pass. Achieves <0.02% accuracy vs GalSim for exponential profiles at all sub-pixel offsets.

Closes #38.

## Review focus

**API — new public interfaces:**
- `kl_pipe/pixel.py` (NEW) — `PixelResponse` ABC with `ft(KX, KY)` and `maxk(threshold)`. `BoxPixel(pixel_scale)` implements 2D sinc FT. Lives on `ImageObs`, default-on via `build_image_obs()`. Pass `pixel_response=None` to disable.
- `kl_pipe/render.py` (NEW) — `RenderConfig` frozen dataclass controlling k-space grid sizing. Factory methods `for_model(model, params, ...)` and `for_priors(model, priors, ...)` compute adaptive oversample/pad from profile × pixel × PSF product scan. `compute_effective_maxk()` is the single source of truth for grid sizing.
- `kl_pipe/model.py` — New abstract methods `IntensityModel.maxk(params, threshold)` and `IntensityModel.stepk(params, folding_threshold)` for adaptive grid sizing. All four intensity models implement these.
- `kl_pipe/observation.py` — `ImageObs` gains `pixel_response` field (pytree-registered). `build_image_obs()` gains `pixel_response` and `render_config` kwargs; default creates `BoxPixel(pixel_scale)`.

**API — changed signatures:**
- `kl_pipe/likelihood.py` — `create_jitted_likelihood_intensity()` and `create_jitted_likelihood_joint()` gain `render_config` / `render_config_int` + `render_config_vel` kwargs, frozen into JIT closure.
- `kl_pipe/model.py` — `Model.render_image()` and `VelocityModel.render_velocity_image()` accept `render_config` via kwargs; adds bin-only fallback when `oversample > 1` but no PSF.
- `kl_pipe/sampling/task.py` — `InferenceTask.from_intensity_obs()` and `from_joint_obs()` compute `RenderConfig.for_priors()` and freeze into likelihood. `_validate_grid_adequacy()` rewritten to use `RenderConfig.for_priors()` (deleted redundant `compute_grid_requirements*` from intensity.py).

**Internals — rendering pipeline (correctness-critical):**
- `kl_pipe/intensity.py` — `_kspace_render_core()` gains `pixel_response` and `render_config` params. Two rendering paths:
  - **Wrap path** (production, `pixel_response is not None`): `profile_FT * sinc` at true k on extended grid → `_wrap_kspace()` folds aliases → IFFT at coarse resolution. Exact pixel integration.
  - **Bin-only path** (`pixel_response=None`, oversample > 1): IFFT at fine resolution → reshape+mean. Compensating phase correction adjusts coarse→fine half-pixel centering before IFFT (fixes centering bug where bin peak was at (31,31) vs wrap at (32,32)).
- `_wrap_kspace()` (NEW) — modular k-space folding, equivalent to GalSim's `kimage._wrap`.
- All four intensity models (`InclinedExponential`, `InclinedSpergel`, `InclinedDeVaucouleurs`, `InclinedSersic`) implement `maxk()`, `stepk()`, `_ft_envelope()` for adaptive grid sizing. Rendering closures pass `pixel_response` and `render_config` through to `_kspace_render_core`.
- Deleted from intensity.py: `compute_effective_maxk` (wrong min-based impl), `compute_grid_requirements` (dead code), `compute_grid_requirements_from_priors` (replaced by `RenderConfig.for_priors`).

**Tests:**
- `tests/test_pixel_integration.py` (NEW, 47 tests) — interface/defaults, point-sampling vs pixel response, convergence (bin-only + sinc+wrap vs GalSim), maxk/stepk validation, grid adequacy, flux conservation, edge cases (sub-pixel galaxy, high-n Sersic), sub-pixel offset stability (4x4 panel diagnostic).
- Test fixture updates across `test_intensity.py`, `test_intensity_sersic.py`, `test_intensity_spergel.py`, `test_likelihood_slices.py`, `test_observation.py`, `test_optimizer_recovery.py`, `test_psf.py` — adapting to `pixel_response` field and `render_config` flow.

**Docs:**
- `docs/psf_pixel_convolution.md` (NEW) — reference doc for k-space rendering pipeline, PixelResponse API, adaptive grid sizing, accuracy characterization.
- Updated: `docs/ROADMAP.md`, `docs/kl_analysis.md`, `docs/psf_image_shape_issue.md` — pixel-specific reference fixes.

## Key design decisions

- **Phase must be applied before wrapping**: aliases at the same wrapped k-bin carry different phases, so the half-pixel correction stays inside `_inclined_sech2_ft`. The bin-only path applies a compensating phase in `_kspace_render_core` (safe because no wrapping occurs).
- **Effective maxk from product scan**: `render.py:compute_effective_maxk()` scans `|profile_FT(k)| * |sinc(k)|` to find the cutoff, rather than taking `min()` of individual maxks. Sinc attenuation roughly halves the needed oversample vs bare profile maxk.
- **RenderConfig frozen for inference**: `InferenceTask` computes worst-case grid from prior bounds at construction time and freezes it into the JIT closure. Grid is deterministic across all MCMC samples.

## Test plan

- [x] 316/316 tests pass (47 new pixel + 269 existing)
- [x] Convergence diagnostic: bin-only converges to GalSim at high N (N=31: <0.1%)
- [x] Sub-pixel offset stability: no position spikes >3x centered residual
- [x] `make test-fast` for full regression

### CODE REVIEW

**Comment by sweverett** [comment]
Here are some of the new diagnostics that i think most clearly show that the code is working.

First, galsim regressions for face-on & inclined disk at different subpixel centers:

subpixel_faceon

subpixel_inclined

These had a max residual vs peak of ~2% before this branch; compare this against the residuals of applying _no pixel response at all_:
point_vs_pixel

This was due to some fundamental mistakes i had previously made in the implementation. The oversampling scheme _was_ converging, but to the wrong values. Now it converges to the correct galsim rendering, albeit much more slowly than the sinc method:
oversample_convergence

Finally, i've tested that our fancy k-space metaparameter predictions perfectly match galsim's internals for some common cases. For example:
maxk_vs_galsim

Sorry for the large PR for what was supposed to be a surgical change. This branch has some core code ideas implemented in `se/inclined-sersic`, so we are merging it there for now.

File: kl_pipe/intensity.py [hunk 1]
```diff
@@ -403,6 +521,35 @@ def _inclined_sech2_ft(
...
+    from kl_pipe.pixel import BoxPixel as _BP
+
+    try:
+        return RenderConfig.for_model(
+            model,
+            params,
+            pixel_scale_f,
+            pixel_response=_BP(pixel_scale_f),
+        )
+    except (KeyError, NotImplementedError, AttributeError):
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/intensity.py]: @sweverett This seems like a useful change.

File: kl_pipe/intensity.py [hunk 2]
```diff
@@ -444,48 +592,63 @@ def _kspace_render_image(
...
                 theta,
                 Nrow * N,
                 Ncol * N,
                 pixel_scale / N,
                 psf_kernel_fft=obs.kspace_psf_fft,
+                pixel_response=obs.pixel_response,
             )
             if N > 1:
                 image = image.reshape(Nrow, N, Ncol, N).mean(axis=(1, 3))
             return image
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/intensity.py]: @sweverett I think this suggestion makes sense as well, would you agree or am I potentially missing something?

File: kl_pipe/intensity.py [hunk 3]
```diff
@@ -1169,45 +1394,60 @@ def render_image(
...
                     theta,
                     Nrow * N,
                     Ncol * N,
                     pixel_scale / N,
                     psf_kernel_fft=obs.kspace_psf_fft,
+                    pixel_response=obs.pixel_response,
                 )
                 if N > 1:
                     image = image.reshape(Nrow, N, Ncol, N).mean(axis=(1, 3))
                 return image
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/intensity.py]: Might need to propagate changes here based on the previous suggestion.

File: kl_pipe/intensity.py
Comment (pranjalrs) [inline kl_pipe/intensity.py:1]: The current file is almost 3000 lines of code and I am wondering whether it is useful to create a module for the intensity model. Separating the routines/class for the fft, Spergel, Sersic, etc into different files would make the code easier to parse/develop.

File: kl_pipe/model.py [hunk 4]
```diff
@@ -150,12 +150,22 @@ def render_image(
...
+                if obs.psf_data is not None:
+                    from kl_pipe.psf import convolve_fft
 
-                return convolve_fft(model_map, obs.psf_data)
+                    return convolve_fft(model_map, obs.psf_data)
+                # no PSF but oversampled → bin for pixel integration
+                N = oversample
+                Nrow = obs.image_pars.Nrow
+                Ncol = obs.image_pars.Ncol
+                return model_map.reshape(Nrow, N, Ncol, N).mean(axis=(1, 3))
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/model.py:168]: @sweverett this might be the problem underlying issue #42.

File: kl_pipe/pixel.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        """
+        ...
+
+    @abstractmethod
+    def maxk(self, threshold: float = 1e-3) -> float:
+        """Wavenumber where the FT amplitude drops below threshold.
+
+        Used for adaptive grid sizing: the effective maxk of a rendering
+        chain is ``min(profile_maxk, pixel_maxk, psf_maxk)``.
+
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/pixel.py:80]: @sweverett double check maxk calculation.

File: kl_pipe/render.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+    k_scan = np.linspace(0.1, profile_maxk, n_scan)
+
+    prof_ft = np.array([model._ft_envelope(k, params) for k in k_scan])
+    sinc_ft = np.abs(np.sinc(k_scan * ps / (2 * np.pi)))
+    product = prof_ft * sinc_ft
+
+    above = k_scan[product > threshold]
+    if len(above) == 0:
+        return float(k_scan[0])
+    return float(above[-1])
```
[the remark this answers is not in the document]
Comment (pranjalrs) [reply kl_pipe/render.py]: @sweverett This suggestion also seems useful.

File: kl_pipe/pixel.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/pixel.py:80]: product is right; fixed the docstring

File: kl_pipe/synthetic.py [hunk 7]
```diff
@@ -294,6 +300,15 @@ def _generate_inclined_kspace_scipy(
...
     I_hat = flux * ft_radial * ft_vertical * phase
 
+    # pixel response: multiply by pixel FT (e.g., sinc for box pixel)
+    if pixel_response is not None:
+        pixel_ft = np.sinc(KX * ps / (2 * np.pi)) * np.sinc(KY * ps / (2 * np.pi))
+        if hasattr(pixel_response, 'pixel_scale'):
+            # use the pixel_response's actual scale (may differ from ps)
+            w = pixel_response.pixel_scale
+            pixel_ft = np.sinc(KX * w / (2 * np.pi)) * np.sinc(KY * w / (2 * np.pi))
+        I_hat = I_hat * pixel_ft
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/synthetic.py]: this was wrong; fixed

File: kl_pipe/intensity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/intensity.py]: fixed

File: kl_pipe/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/model.py:168]: it was; fixed

File: kl_pipe/intensity.py
Comment (sweverett) [reply kl_pipe/intensity.py:1]: This is good feedback but i will leave for a separate & contained PR

**Comment by sweverett** [comment]
Closes #42
